e2e/scripts/common.func

```diff
@@ -0,0 +1,178 @@
+}
+
+interactive-ctrl-c() {
+    _check-script
+    interactive-send $'\x03'
+}
+
+interactive-ctrl-d() {
+    _check-script
+    interactive-send $'\x04'
+}
+
+interactive-eof() {
+    _check-script
+    interactive-expect "<EOF>"
+}
+
+exit-interactive() {
+    if [ $# != 1 ]; then
+        fatal "exit-interactive need an exit code argument"
+    fi
+    if [ ! -f ${_script} ]; then
+        fatal "no interactive script provided"
+    fi
+    if [ ! -f ${_cmdline} ]; then
+        fatal "no command line provided in start-interactive"
+    fi
+    expect-exit-interactive $1 `cat ${_cmdline}`
+    rm -f ${_script}
+    rm -f ${_cmdline}
+}
+
+# wrapper around default test builtin
+expect-failure() {
+    expect-exit 1 "$@"
+}
+
+expect-success() {
+    expect-exit 0 "$@"
+}
+
+stdout-contains() {
+    expect-search output "$@"
+}
+
+stderr-contains() {
+    expect-search error "$@"
+}
+
+test-match() {
+    expect-exit 0 $1 /bin/sh -c "[ \"$2\" = \"$3\" ]"
+}
+
+# manage test file/directory
+mktmpdir() {
+    create-tmpdir ${TESTDIR}
+}
+
+mktmpfile() {
+    create-tmpfile ${TESTDIR}
+}
+
+# test image
+get-test-image() {
+    image="${TESTDIR}/test.sif"
+
+    if [ ! -f "${image}" ]; then
+        definition="${SOURCEDIR}/e2e/testdata/Singularity"
+        sudo singularity build -F ${image} ${definition} >${TESTDIR}/image.out 2>&1
+        if [ $? != 0 ]; then
+            output=`cat ${TESTDIR}/image.out`
+            fatal "failed to build image ${image}: ${output}"
+        fi
+    fi
+
+    echo ${image}
+}
+
+verify-image() {
+    if [ $? = 0 ]; then
+        name="$1"
+        image="$2"
+        expect-exit 1 "${name}/False"        singularity exec ${image} false
+        expect-exit 0 "${name}/RunScript"    singularity exec ${image} test -f /.singularity.d/runscript
+        expect-exit 0 "${name}/OneBase"      singularity exec ${image} test -f /.singularity.d/env/01-base.sh
+        expect-exit 0 "${name}/ActionsShell" singularity exec ${image} test -f /.singularity.d/actions/shell
+        expect-exit 0 "${name}/ActionsExec"  singularity exec ${image} test -f /.singularity.d/actions/exec
+        expect-exit 0 "${name}/ActionsRun"   singularity exec ${image} test -f /.singularity.d/actions/run
+        expect-exit 0 "${name}/Environment"  singularity exec ${image} test -L /environment
+        expect-exit 0 "${name}/Singularity"  singularity exec ${image} test -L /singularity
+    fi
+}
+
+check-command() {
+    if ! command -v $1 >/dev/null 2>&1; then
+        test-skip-script "$1 command not found in PATH"
+    fi
+}
+
+check-overlayfs() {
+    sudo modprobe overlay >/dev/null 2>&1
+    check-fs overlay
+    if [ $? != 0 ]; then
+        test-skip-script "overlay filesystem seems not supported"
+    fi
+}
+
+run-command() {
+    $@ >${TESTDIR}/run-command.out 2>&1
+    if [ $? != 0 ]; then
+        output=`cat ${TESTDIR}/run-command.out`
+        fatal "$@ failed: ${output}"
+    fi
+}
+
+check-kernel-version() {
```

Comment:
Using the kernel version for a check on a feature usually doesn't work since a lot of distros add patches to the kernel out of order. I've learned the hard way relying on the kernel version is usually a problem. Best to check if something works instead.

Reply:
Agreed